apis/python/src/tiledb/vector_search/object_api/object_index.py

```diff
@@ -0,0 +1,390 @@
+import base64
+from collections import OrderedDict
+from typing import Any, Dict, Mapping, Optional
+
+import cloudpickle
+import numpy as np
+
+import tiledb
+import tiledb.vector_search.object_api as object_api
+from tiledb.cloud.dag import Mode
+from tiledb.vector_search import FlatIndex
+from tiledb.vector_search import IVFFlatIndex
+from tiledb.vector_search import flat_index
+from tiledb.vector_search import ingest
+from tiledb.vector_search import ivf_flat_index
+from tiledb.vector_search.embeddings import ObjectEmbedding
+from tiledb.vector_search.object_readers import ObjectReader
+from tiledb.vector_search.storage_formats import STORAGE_VERSION
+from tiledb.vector_search.storage_formats import storage_formats
+
+TILEDB_CLOUD_PROTOCOL = 4
+
+
+class ObjectIndex:
+    def __init__(
+        self,
+        uri: str,
+        config: Optional[Mapping[str, Any]] = None,
+        timestamp=None,
+        load_embedding: bool = True,
+        **kwargs,
+    ):
+        with tiledb.scope_ctx(ctx_or_config=config):
+            self.uri = uri
+            self.config = config
+            self.timestamp = timestamp
+            group = tiledb.Group(uri, "r")
+            self.index_type = group.meta["index_type"]
+            group.close()
+            if self.index_type == "FLAT":
+                self.index = FlatIndex(
+                    uri=uri, config=config, timestamp=timestamp, **kwargs
+                )
+            elif self.index_type == "IVF_FLAT":
+                self.index = IVFFlatIndex(
+                    uri=uri, config=config, timestamp=timestamp, **kwargs
+                )
+
+            self.object_reader_source_code = self.index.group.meta[
+                "object_reader_source_code"
+            ]
+            self.object_reader_class_name = self.index.group.meta[
+                "object_reader_class_name"
+            ]
+            self.object_reader_kwargs = cloudpickle.loads(
```

Comment:
See the previous comment for a detailed explanation. 

Since we package the class code and we expect most `init_kwargs` to be primitive types there shouldn't be many problems with Python versions.

The object API is only available in Python for now. We will need to design how these concepts(serialization, remote execution) can work in C++. A similar object API that performs local execution with code existing in the same env would be easy to replicate though.

Reply:
I strongly believe this design needs to be considered up front. We have to maintain compatibility, so by introducing it was pickled objects we have to support this going forward. We need to consider the medium and long term objectives when we introduce new on-disk structures. We haven't done a release yet so lets have serious consideration on how we can remove the pickling so we aren't locking all datasets to limited python versions.

Reply:
Some comments on Python version locking (this will not require a dataset Pyrhon version lock):
- The code part is serialized as a string so there is no version incompatibility involved. This is similar to the UDF function code that we can register on cloud.
- The `init_kwargs` are pickled atm. If there is a concern with version locking we can restrict them to primitive types and even serialize as json or other format.